File: scripts/pick_and_place.py
import rospy
import moveit_commander
from geometry_msgs.msg import Pose, Point, Quaternion, Vector3
import moveit_msgs.msg
from tf.transformations import quaternion_from_euler, euler_matrix,translation_matrix
import shape_msgs.msg
import sys
import math
import visualization_msgs.msg 
import std_msgs.msg
import numpy as np

# Constants for opening and closing the gripper
OPEN_GRIPPER = [0.035, 0.035]
CLOSE_GRIPPER = [0.011, 0.011]
end_effector_palm_length = 0.058 * 1.8 # 1.4 is padding
# box1 parameters
box1_height = 0.2
box1_xposition = 0.6
box1_yposition = 0.2
# box2 parameters
box2_height = 0.1
box2_xposition = 0
box2_yposition = 0.6
# rod parameters
rod_height = 0.2

class PickAndPlace(object):
    """Pick and place a box using moveit_commander"""

    def __init__(self):
        super(PickAndPlace, self).__init__()

        # Initialize moveit_commander and ROS node
        moveit_commander.roscpp_initialize(sys.argv)
        rospy.init_node('my_interfaces', anonymous=True)

        self.pose_point_pub = rospy.Publisher('pose_point', visualization_msgs.msg.Marker, queue_size=10)
        # Create a RobotCommander object
        robot = moveit_commander.RobotCommander()
        # Create a PlanningSceneInterface object
        scene = moveit_commander.PlanningSceneInterface()
        # Create a MoveGroupCommander object for the panda arm
        arm = moveit_commander.MoveGroupCommander("panda_arm")
        # Set the number of planning attempts to 120
        arm.set_num_planning_attempts(120)
        # Set the planning time to 10 seconds
        arm.set_planning_time(10)
        # End effector group
        end_effector = moveit_commander.MoveGroupCommander("panda_hand")
        # Get the reference frame for the robot
        planning_frame = arm.get_planning_frame()
        print("============ Planning frame: %s" % planning_frame)
        # Get the end-effector link for the group
        eef_link = arm.get_end_effector_link()
        print("============ End effector link: %s" % eef_link)
        # Get a list of all the groups in the robot
        group_names = robot.get_group_names()
        print("============ Available Planning Groups:", robot.get_group_names())
        # Print the entire state of the robot for debugging purposes
        print("============ Printing robot state")
        print(robot.get_current_state())
        print("")

        self.robot = robot
        self.scene = scene
        self.arm = arm
        self.end_effector = end_effector
        self.planning_frame = planning_frame
        self.eef_link = eef_link
        self.group_names = group_names


    def create_collision_scene(self) -> None:
        """Create a collision scene"""
        # floor limit 
        floor = self.create_collision_object("floor", [2.5, 2.5, 0.01], [0, 0, -0.01], 0)
        self.scene.add_object(floor)
        box1 = self.create_collision_object("box1", [0.2, 0.4, box1_height], [box1_xposition, box1_yposition, box1_height/2], 0)
        self.scene.add_object(box1)
        box2 = self.create_collision_object("box2", [0.3, 0.2, box2_height], [box2_xposition, box2_yposition, box2_height/2], 0)
        self.scene.add_object(box2)
        rod = self.create_collision_object("rod", [0.02, 0.02, rod_height], [box1_xposition, box1_yposition, rod_height/2 + box1_height], 45)
        self.scene.add_object(rod)


    def get_joint_values(self):
        """Get the starting joint values"""
        return self.arm.get_current_joint_values()


    def go_to_joint_state(self, joint_goal) -> None:
        """Go to a specified joint state"""
        self.arm.go(joint_goal, wait=True)
        self.arm.stop()
    

    def close_gripper(self) -> None:
        """Close the gripper"""
        self.end_effector.set_named_target("close")
        self.end_effector.go()


    def go_to_pose_target(self, translation:list = None, rotation:list = None, relative:bool = False) -> None:
        """Go to a specified pose, the pose is selected by the user as the desired location
        of the palm of the hand. The orientation is specified by the user as a list of Euler angles
        and the rotation in which they are applied (static or relative needs to be specified in the 
        axis=<order> flag of the euler tf functions)."""

        if relative is False:
            # Set the target orientation initalize position and orientation messages
            rotation_rads = [math.radians(rotation[0]), math.radians(rotation[1]), math.radians(rotation[2])]
            homogeneous_mat_arm = euler_matrix(rotation_rads[0], rotation_rads[1], rotation_rads[2], axes='szyz')
            # add the translation to the homogeneous transformation matrix 
            homogeneous_mat_arm[:3, 3] = translation[:3]     
            # create a homogeneous transformation matrix for the end effector with no rotation
            homogeneous_trans_end_effector = translation_matrix([0, 0, end_effector_palm_length])
            # multiply the homogeneous transformation matrix of the arm by the inverse of the homogeneous transformation matrix of the end effector
            homogeneous_mat = np.dot(homogeneous_mat_arm, np.linalg.inv(homogeneous_trans_end_effector))
            # quaternion_from_euler using input rotation values 
            quaternion = quaternion_from_euler(rotation_rads[0], rotation_rads[1], rotation_rads[2], axes='szyz')           
            # create message types for the pose target
            orientation = Quaternion(quaternion[0], quaternion[1], quaternion[2], quaternion[3])
            position = Point(homogeneous_mat[0,3], homogeneous_mat[1,3], homogeneous_mat[2,3])
            # Set the target pose message
            pose_target = Pose(position, orientation)
            self.add_pose_arrow(position, rotation_rads[2])
            input("Pose Target Set, Press Enter to Continue...")
            self.arm.set_pose_target(pose_target)


        # Needs Debugging - Path Planning Error
        else:
            # Set the target orientation initalize position and orientation messages
            rotation = [math.radians(rotation[0]), math.radians(rotation[1]), math.radians(rotation[2])]
            orientation = quaternion_from_euler(rotation[0], rotation[1], rotation[2], axes='rzyz')
            orientation = Quaternion(orientation[0], orientation[1], orientation[2], orientation[3])
            position = Point(translation[0], translation[1], translation[2])
            # Set the target pose message
            pose_target = Pose(position, orientation)
            self.arm.set_pose_target(pose_target, end_effector_link="panda_link8")
            
        self.arm.set_goal_tolerance(0.001)
        self.arm.go(wait=True)
        self.arm.stop()
        # Clear the target
        self.arm.clear_pose_targets()   

        
    def create_collision_object(self, id, dimensions:list, position:list, rotation_z:int) -> moveit_msgs.msg.CollisionObject:
        """Create collision objects and return it."""
        object = moveit_msgs.msg.CollisionObject()
        object.header.frame_id = self.planning_frame # "panda_link0" satisfies the std_msgs.msg.Header requirements
        object.id = id
    
        solid = shape_msgs.msg.SolidPrimitive()
        solid.type = solid.BOX
        solid.dimensions = dimensions
        object.primitives = [solid]

        # rotation about the z axis 
        rotation = [0, 0, math.radians(rotation_z)]
        quaternion = quaternion_from_euler(rotation[0], rotation[1], rotation[2], axes='szyz')

        # Position messages initailized and assigned
        object_point = Point(position[0], position[1], position[2])
        orientation = Quaternion(quaternion[0], quaternion[1], quaternion[2], quaternion[3])
        object_pose = Pose(object_point, orientation)

        object.primitive_poses = [object_pose]
        object.operation = object.ADD
        return object

    # attach the rod collision object to grasping group
    def attach_rod(self) -> None:
        """Attach an object to the robot"""
        grasping_group = "panda_hand"
        touch_links = self.robot.get_link_names(group=grasping_group)
        self.scene.attach_box(self.eef_link, "rod", touch_links=touch_links)


    def get_rod_position(self) -> None:
        """Get the rod position"""
        # returns a list of the rod position
        rod_position = self.scene.get_object_poses(["rod"])["rod"].position
        # convert geometry_msgs.msg.Point to list
        return [rod_position.x, rod_position.y, rod_position.z]


    def clean_scene(self) -> None:
        """Clean the scene"""
        self.scene.remove_world_object("floor")
        self.scene.remove_world_object("box1")
        self.scene.remove_world_object("box2")
        self.scene.remove_world_object("rod")


    def close_gripper(self) -> None:
        """Close the gripper"""
        # using the CLOSE_GRIPPER variable set the panda_finger_joint1 and panda_finger_joint2 to the desired value
        print(f"Current End Effector Joint Values (Closed): {self.end_effector.get_current_joint_values()}")
        self.end_effector.set_joint_value_target(CLOSE_GRIPPER)
        self.end_effector.go(wait=True)
        

    def open_gripper(self) -> None:
        """Open the gripper"""
        # using the OPEN_GRIPPER variable set the panda_finger_joint1 and panda_finger_joint2 to the desired value
        print(f"Current End Effector Joint Values (Open): {self.end_effector.get_current_joint_values()}")
        self.end_effector.set_joint_value_target(OPEN_GRIPPER)
        self.end_effector.go(wait=True)


    def add_pose_arrow(self, desired_position:Point, z_rotation:float) -> None:
        """Publish a marker at the desired pose"""    

        # Set up the marker message- edit the marker to display an arrow given orientation and positon
        marker = visualization_msgs.msg.Marker()
        marker.ns = "arrow"
        marker.id = 0
        marker.header.frame_id = 'panda_link0'
        marker.type = marker.ARROW
        marker.action = marker.ADD
        marker.scale = Vector3(end_effector_palm_length, 0.02, 0.02)
        marker.color = std_msgs.msg.ColorRGBA(1.0, 0.0, 1.0, 1.0)

        quaternion = quaternion_from_euler(0, 0, z_rotation, axes='szyz')
        desired_pose = Pose(desired_position, Quaternion(quaternion[0], quaternion[1], quaternion[2], quaternion[3]))

        marker.pose = desired_pose
        # Publish the marker
        self.pose_point_pub.publish(marker)

    def remove_pose_arrow(self) -> None:
        """Remove the marker from the scene"""
        marker = visualization_msgs.msg.Marker()
        marker.ns = "arrow"
        marker.id = 0
        marker.action = marker.DELETE
        self.pose_point_pub.publish(marker)

    def move_home(self, home_joint_pos) -> None:
        """Move the robot to the home position"""
        self.go_to_joint_state(home_joint_pos)


    def pick_rod(self) -> None:
        rod_position = self.get_rod_position()
        self.go_to_pose_target([round(rod_position[0],2), round(rod_position[1],2), round(rod_position[2],2)], [45, 90, 45], relative=False)
        input("[PROGRESS] Pose Reached, Press any key to close the gripper")
        self.close_gripper()
        # attach the rod to the end effector
        self.attach_rod()
        self.remove_pose_arrow()
    
    def place_rod(self) -> None:
        place_point_position = [box2_xposition, box2_yposition, rod_height/2 + box2_height]
        self.go_to_pose_target(place_point_position, [45, 90, 90], relative=False)
        self.remove_pose_arrow()
        self.open_gripper()
        # detach rod from end effector
        self.scene.remove_attached_object(self.eef_link, name="rod")



def main(empty=False):
    try:
        # initialize the node
        pick_and_place = PickAndPlace()
        # retrieve critical parameters and set up the scene
        home_joint_pos = pick_and_place.get_joint_values()
        print(f"Joint Values: {home_joint_pos}")

        if empty is False:
            print("[INFO] Setting up scene with Collision Objects")
            pick_and_place.create_collision_scene()
            input("[PROGRESS] Scene Set Up and Initial Pose Identified, Press any key to Pick up the Rod")
            # send the panda arm to the desired pose
            pick_and_place.pick_rod()
            input("[PROGRESS] Rod Picked, Press any key to place")
            pick_and_place.place_rod()
            input("[PROGRESS] Rod Placed, Press any key to move to home position")
            pick_and_place.move_home(home_joint_pos)
            input("[PROGRESS] Home Position Reached, Press any key to open the gripper and remove the objects from the scene")
            pick_and_place.clean_scene()

        else:
            print("[INFO] Setting up Empty Scene")
            '''For testing purposes, you can set up the scene without any collision objects'''
            # pick_and_place.add_pose_arrow([0.59, 0.0, 0.45], [0, 90, 45])
            # input("[PROGRESS] Scene Set Up and Initial Pose Identified, Press any key to move to desired pose")
            # # send the panda arm to the desired pose
            # pick_and_place.go_to_pose_target([0.59, 0.0, 0.45], [0, 90, 45])
            # input("[PROGRESS] Pose Reached, Press any key to close the gripper/move to home position")
            # #pick_and_place.close_gripper()
            #input("[PROGRESS] Gripper Closed, Press any key to move to home position")
            
            pick_and_place.go_to_pose_target([0.6, 0.2, 0.6], [45, 90, 45], relative=False) 
            input("[PROGRESS] Static Pose Reached, Press any key to continue")
            
            
            # try a relative move
            #pick_and_place.go_to_pose_target([0.0, 0.0, 0.05], [0, 0, 0], relative=True)
            #input("[PROGRESS] Relative Pose Reached, Press any key to continue")

            pick_and_place.move_home(home_joint_pos)
            pick_and_place.remove_pose_arrow()

    
    except rospy.ROSInterruptException:
        pass
        
if __name__ == "__main__":
    main(empty=False)
        

File: scripts/pick_and_place.cpp
// ROS
#include <ros/ros.h>
#include <cmath>
#include <moveit/move_group_interface/move_group_interface.h>
#include <moveit/planning_scene_interface/planning_scene_interface.h>
#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
#include <tf2/LinearMath/Quaternion.h>
#include <visualization_msgs/Marker.h> 
#include <Eigen/Geometry>
#include <geometry_msgs/Quaternion.h>
#include <geometry_msgs/Point.h>
#include <tf2_eigen/tf2_eigen.h>
#include <sensor_msgs/JointState.h>
#include <geometry_msgs/Pose.h>

namespace pnp
{
    // This class contains all the functions needed to perform the pick and place operation
class PickandPlace{
private:

    ros::Publisher pose_point_pub;
    // Set global parameters of panda arm
    const std::vector<double> OPEN_GRIPPER = {0.035, 0.035};
    const std::vector<double> CLOSE_GRIPPER = {0.011, 0.011};
    const double end_effector_palm_length = 0.058 * 1.8; // 1.4 is padding

    // Map is the python equivalent of a dictionary
    const std::map<std::string, double> box1 = {
        {"x_pos", 0.6}, {"y_pos", 0.2}, {"z_height", 0.2}
    };
    const std::map<std::string, double> box2 = {
        {"x_pos", 0}, {"y_pos", 0.6}, {"z_height", 0.1}
    };
    const double rod_height = 0.2;
    
    // MoveIt operates on sets of joints called "planning groups" and stores them in an object called
    // the `JointModelGroup`. Throughout MoveIt the terms "planning group" and "joint model group"
    // are used interchangeably.
    const std::string PLANNING_GROUP_ARM = "panda_arm";
    const std::string PLANNING_GROUP_GRIPPER = "panda_hand";

    moveit::planning_interface::MoveGroupInterface move_group_interface_arm;
    moveit::planning_interface::MoveGroupInterface move_group_interface_gripper;


    // Create a moveit::planning_interface::MoveGroupInterface::Plan object to store the movements
    moveit::planning_interface::MoveGroupInterface::Plan plan;

    // planning_scene_interface allows us to add and remove collision objects in the world
    moveit::planning_interface::PlanningSceneInterface planning_scene_interface;

    const robot_state::JointModelGroup* joint_model_group_arm;
    const robot_state::JointModelGroup* joint_model_group_gripper;

    std::vector<double> floor_dimensions = {2.5, 2.5, 0.01};
    std::vector<double> floor_position = {0.0, 0.0, -0.01};

    std::vector<double> box1_dimensions = {0.2, 0.4, box1.at("z_height")};
    std::vector<double> box1_position = {box1.at("x_pos"), box1.at("y_pos"), box1.at("z_height")/2.0};

    // empty vector to be filled with the joint values of the robot
    std::vector<double> home_joint_values;




public:
    // The init function
    PickandPlace(ros::NodeHandle nh) : move_group_interface_arm(PLANNING_GROUP_ARM), move_group_interface_gripper(PLANNING_GROUP_GRIPPER)
    {  
        pose_point_pub = nh.advertise<visualization_msgs::Marker>("pose_point", 10);

        // ROS spinning must be running for the MoveGroupInterface to get information
        // about the robot's state. One way to do this is to start an AsyncSpinner
        // beforehand.
        ros::AsyncSpinner spinner(1);
        spinner.start();
        
        // set arm planning time 
        const double PLANNING_TIME = 5.0;
        move_group_interface_arm.setPlanningTime(PLANNING_TIME);

        // Raw pointers are frequently used to refer to the planning group for improved performance.
        joint_model_group_arm = move_group_interface_arm.getCurrentState()->getJointModelGroup(PLANNING_GROUP_ARM);
        joint_model_group_gripper = move_group_interface_gripper.getCurrentState()->getJointModelGroup(PLANNING_GROUP_GRIPPER);
    }

    void writeRobotDetails()
    {
        // Print information about the robot
        ROS_INFO_NAMED("pnp", "Planning frame: %s", move_group_interface_arm.getPlanningFrame().c_str());

        std::vector<std::string> linkNames = move_group_interface_arm.getLinkNames();
        std::string linkNamesArm;
        for (const auto& link : linkNames)
        {
            linkNamesArm += link + ", ";
        }
        ROS_INFO_NAMED("pnp", "Arm links: %s", linkNamesArm.c_str());

        // ROS_INFO_NAMED the arm joint names
        std::vector<std::string> jointNamesArm = move_group_interface_arm.getJoints();
        std::string jointNamesArmString;
        for (const auto& joint : jointNamesArm)
        {
            jointNamesArmString += joint + ", ";
        }
        ROS_INFO_NAMED("pnp", "Arm joint names: %s", jointNamesArmString.c_str());

        // ROS_INFO_NAMED the gripper joint names
        std::vector<std::string> jointNamesGripper = move_group_interface_gripper.getJoints();
        std::string jointNamesGripperString;
        for (const auto& joint : jointNamesGripper)
        {
            jointNamesGripperString += joint + ", ";
        }

        ROS_INFO_NAMED("pnp", "Gripper joints names: %s", jointNamesGripperString.c_str());

        // We can get a list of all the groups in the robot:
        std::vector<std::string> planningGroups = move_group_interface_arm.getJointModelGroupNames();
        std::string planningGroupsString;
        for (const auto& groups : planningGroups)
        {
            planningGroupsString += groups + ", ";
        }
        ROS_INFO_NAMED("pnp", "Available Planning Groups: %s", planningGroupsString.c_str());
        // Add a delay before calling getCurrentJointValues()
        ros::Duration(1.0).sleep();
        
        // using getJointStates to get the current joint values put in a block to print the error 
        home_joint_values = move_group_interface_arm.getCurrentJointValues();
        std::string jointValuesString;
        for (const auto& joint : home_joint_values)
        {
            jointValuesString += std::to_string(joint) + ", ";
        }
        ROS_INFO_NAMED("pnp", "Current joint values: %s", jointValuesString.c_str());
        
        // add a sleep 
        //ros::Duration(1000.0).sleep();
        

    }

    // std::vector<double> jointStateCallback(const sensor_msgs::JointState::ConstPtr& msg, const robot_state::JointModelGroup* joint_model_group_arm)
    // {
    //     // Get the joint values for the arm
    //     std::vector<double> arm_joint_values;
    //     for (int i = 0; i < msg->name.size(); i++)
    //     {
    //         if (joint_model_group_arm->hasJointModel(msg->name[i]))
    //         {
    //         arm_joint_values.push_back(msg->position[i]);
    //         }
    //     }

    //     return arm_joint_values;
    // }

    // std::vector<double> getArmJointStatesFromTopic(const robot_state::JointModelGroup* joint_model_group_arm)
    // {   
    //     // Create a node handle
    //     ros::NodeHandle nh;
    //     // Subscribe to the joint state topic
    //     sensor_msgs::JointState::ConstPtr msg;
    //     ros::Subscriber sub = nh.subscribe<sensor_msgs::JointState>("/joint_states", 1, [&](const sensor_msgs::JointState::ConstPtr& joint_state_msg) {
    //         msg = joint_state_msg;
    //     });

    //     // Spin the node
    //     ros::spinOnce();



    //     //Return the joint values
    //     return jointStateCallback(msg, joint_model_group_arm);
    // }

        
    void createCollisionObject(std::string id, std::vector<double> dimensions, std::vector<double> position, double rotation_z)
    {
        // Create collision object
        moveit_msgs::CollisionObject collision_object;
        collision_object.header.frame_id = "panda_link0";
        collision_object.id = id;

        // Convert rotation to quaternion
        double rotation_radians = rotation_z * M_PI / 180.0;
        tf2::Quaternion quaternion;
        quaternion.setRPY(0.0, 0.0, rotation_radians);

        // Define a box to add to the world.
        shape_msgs::SolidPrimitive primitive;
        primitive.type = primitive.BOX;
        primitive.dimensions.resize(3);
        primitive.dimensions[primitive.BOX_X] = dimensions[0];
        primitive.dimensions[primitive.BOX_Y] = dimensions[1];
        primitive.dimensions[primitive.BOX_Z] = dimensions[2];

        // Define a pose for the box (specified relative to frame_id)
        geometry_msgs::Pose box_pose;
        box_pose.position.x = position[0];
        box_pose.position.y = position[1];
        box_pose.position.z = position[2];
        box_pose.orientation.x = quaternion.x();
        box_pose.orientation.y = quaternion.y();
        box_pose.orientation.z = quaternion.z();
        box_pose.orientation.w = quaternion.w();

        collision_object.primitives.push_back(primitive);
        collision_object.primitive_poses.push_back(box_pose);
        collision_object.operation = collision_object.ADD;

        std::vector<moveit_msgs::CollisionObject> collision_objects;
        collision_objects.push_back(collision_object);

        // Now, let's add the collision object into the world
        // (using a vector that could contain additional objects)
        planning_scene_interface.addCollisionObjects(collision_objects);

        // ROS_INFO_NAMED added object frame.id to world
        ROS_INFO_NAMED("pnp", "Added %s into the world", id.c_str());
    }

    void createCollisionScene()
    {
        // Floor
        std::vector<double> floor_dimensions = {2.5, 2.5, 0.01};
        std::vector<double> floor_position = {0.0, 0.0, -0.01};
        createCollisionObject("floor", floor_dimensions, floor_position, 0.0);

        // Box 1
        std::vector<double> box1_dimensions = {0.2, 0.4, box1.at("z_height")};
        std::vector<double> box1_position = {box1.at("x_pos"), box1.at("y_pos"), box1.at("z_height") / 2.0};
        createCollisionObject("box1", box1_dimensions, box1_position, 0.0);

        // Box 2
        std::vector<double> box2_dimensions = {0.3, 0.2, box2.at("z_height")};
        std::vector<double> box2_position = {box2.at("x_pos"), box2.at("y_pos"), box2.at("z_height") / 2};
        createCollisionObject("box2", box2_dimensions, box2_position, 0.0);

        // Rod
        std::vector<double> rod_dimensions = {0.02, 0.02, rod_height};
        std::vector<double> rod_position = {box1.at("x_pos"), box1.at("y_pos"), rod_height / 2.0 + box1.at("z_height")};
        createCollisionObject("rod", rod_dimensions, rod_position, 45.0);
    }

    void clean_scene() {
        std::vector<std::string> object_ids;
        object_ids.push_back("floor");
        object_ids.push_back("box1");
        object_ids.push_back("box2");
        object_ids.push_back("rod");
        planning_scene_interface.removeCollisionObjects(object_ids);
    }


    void set_pose_target(std::vector<double> translation, std::vector<double> rotation)
    {
        // Euler angles to radians
        std::vector<double> rotation_rads = {
            rotation[0] * M_PI / 180.0,
            rotation[1] * M_PI / 180.0,
            rotation[2] * M_PI / 180.0
        };

        Eigen::Matrix4d homogeneous_mat_arm = Eigen::Matrix4d::Identity();
        Eigen::Matrix3d Rz1, Ry, Rz2;
        Rz1 = (Eigen::AngleAxisd(rotation_rads[0], Eigen::Vector3d::UnitZ())).toRotationMatrix();
        Ry = (Eigen::AngleAxisd(rotation_rads[1], Eigen::Vector3d::UnitY())).toRotationMatrix();
        Rz2 = (Eigen::AngleAxisd(rotation_rads[2], Eigen::Vector3d::UnitZ())).toRotationMatrix();
        Eigen::Matrix3d R = Rz1 * Ry * Rz2;
        homogeneous_mat_arm.block<3, 3>(0, 0) = R;
        homogeneous_mat_arm(0, 3) = translation[0];
        homogeneous_mat_arm(1, 3) = translation[1];
        homogeneous_mat_arm(2, 3) = translation[2];

        // Create a homogeneous transformation matrix for the end effector with no rotation
        Eigen::Matrix4d homogeneous_trans_end_effector = Eigen::Matrix4d::Identity();
        homogeneous_trans_end_effector(2, 3) = end_effector_palm_length;

        // Multiply the homogeneous transformation matrix of the arm by the inverse of the homogeneous transformation matrix of the end effector
        Eigen::Matrix4d homogeneous_mat = homogeneous_mat_arm * homogeneous_trans_end_effector.inverse();

        // Create a quaternion from euler angles
        tf2::Quaternion quaternion;
        // get a quaternion from the rotation matrix R
        Eigen::Quaterniond eigen_quat(R);
        // convert eigen quaternion to tf quaternion
        tf2::convert(tf2::toMsg(eigen_quat), quaternion);

        // Create message types for the pose target
        geometry_msgs::Quaternion orientation;
        orientation.x = quaternion.x();
        orientation.y = quaternion.y();
        orientation.z = quaternion.z();
        orientation.w = quaternion.w();

        geometry_msgs::Point position;
        position.x = homogeneous_mat(0, 3);
        position.y = homogeneous_mat(1, 3);
        position.z = homogeneous_mat(2, 3);

        // Set the target pose message
        geometry_msgs::Pose pose_target;
        pose_target.position = position;
        pose_target.orientation = orientation;


        // add pose arrow
        add_pose_arrow(pose_target.position, rotation_rads[2]);

        // set the target pose
        move_group_interface_arm.setPoseTarget(pose_target);

        // print target pose successfully set
        ROS_INFO_NAMED("pnp", "Pose target successfully set");

        // move to arm to the target pose
        bool success = (move_group_interface_arm.plan(plan) == moveit::planning_interface::MoveItErrorCode::SUCCESS);

        // print if the arm was able to move to the target pose
        ROS_INFO_NAMED("pnp", "Visualizing plan 1 (pose goal) %s", success ? "" : "FAILED");

        // execute the plan
        move_group_interface_arm.move();

    }

    void add_pose_arrow(geometry_msgs::Point desired_position, float z_rotation) {
        // Publish a marker at the desired pose
        visualization_msgs::Marker marker;
        marker.ns = "arrow";
        marker.id = 0;
        marker.header.frame_id = "panda_link0";
        marker.type = visualization_msgs::Marker::ARROW;
        marker.action = visualization_msgs::Marker::ADD;
        marker.scale.x = end_effector_palm_length;
        marker.scale.y = 0.02;
        marker.scale.z = 0.02;
        marker.color.r = 1.0;
        marker.color.g = 0.0;
        marker.color.b = 1.0;
        marker.color.a = 1.0;

        tf2::Quaternion quaternion;
        quaternion.setRPY(0, 0, z_rotation);
        geometry_msgs::Pose Pose;
        Pose.position = desired_position;
        Pose.orientation = tf2::toMsg(quaternion);
        marker.pose = Pose;
        
        // Publish the marker
        pose_point_pub.publish(marker);

        // print pose arrow successfully published
        ROS_INFO_NAMED("pnp", "Pose arrow successfully published");

    }

    std::vector<double> get_rod_position() 
    {
        std::vector<std::string> object_ids;
        object_ids.push_back("rod");
        // type std::map<std::string, geometry_msgs::Pose>.
        auto object_poses = planning_scene_interface.getObjectPoses(object_ids);
        // Check if "rod" exists in the returned map
        auto search = object_poses.find("rod");
        if (search != object_poses.end()) {
            auto rod_pose = search->second;
            // convert geometry_msgs::Pose to vector
            return {rod_pose.position.x, rod_pose.position.y, rod_pose.position.z};
        } else {
            throw std::runtime_error("Rod not found in object_poses");
        }
    }


    void pick(){
        
    }


    void run()
    {   
        // Write robot details
        writeRobotDetails();

        // Create collision scene
        createCollisionScene();

        // get rod position
        std::vector<double> rod_position = get_rod_position();
        // print rod position
        ROS_INFO_NAMED("pnp", "Rod position: %f, %f, %f", rod_position[0], rod_position[1], rod_position[2]);

        ros::Duration(2.0).sleep();
    }
};
}

int main(int argc, char** argv)
{
    ros::init(argc, argv, "pick_and_place");
    ros::NodeHandle nh;
    // add a short sleep so the node can finish initializing
    ros::Duration(0.5).sleep();

    // Instantiate the PickandPlace class
    pnp::PickandPlace pnp(nh);

    // Run pick and place operations
    pnp.run();

    // Shutdown the node and join the thread back before exiting
    ros::shutdown();

    return 0;
}


